### PR TITLE
Epoch size default behavior

### PR DESCRIPTION
**Pull request by snarayan21**
## Description of changes:
Makes sure we sample from streams in proportion with the number of samples per stream when none of proportion, repeat, and choose are specified, but epoch size is specified. Builds on previous PR dealing with epoch size for single streams.

## Issue #, if available:
none

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [x] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [x] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#submitting-a-contribution))
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

**Review by karan6181 — COMMENTED** [review]
Can you add the below two lines in `.gitignore` file to ensure DS_Store files are not getting committed by mistakenly ?

```
# OS X
.DS_Store
```
Reviewing at the other section of your PR.

File: tests/test_streaming.py [hunk 1]
```diff
@@ -58,6 +59,94 @@ def test_dataloader_epoch_size_no_streams(local_remote_dir: Tuple[str,
...
+    samples_seen_stream2 = 0
+    samples_seen = 0
+    for batch in dataloader:
+        samples = batch['sample']
+        samples_seen += samples.size(dim=0)
+        stream1_seen = (samples < 600).sum().item()
+        stream2_seen = (samples > 600).sum().item()
+        samples_seen_stream1 += stream1_seen
+        samples_seen_stream2 += stream2_seen
+        print(samples)
```
Comment (karan6181) [inline tests/test_streaming.py:120]: Maybe you missed removing `print` statement ?

File: tests/test_streaming.py [hunk 2]
```diff
@@ -58,6 +59,94 @@ def test_dataloader_epoch_size_no_streams(local_remote_dir: Tuple[str,
...
+        print(stream1_seen)
+        print(stream2_seen)
+
+    print('FINAL TOTAL')
+    print(samples_seen)
+    print(samples_seen_stream1)
+    print(samples_seen_stream2)
+
+    if epoch_size % num_canonical_nodes != 0:
```
Comment (karan6181) [inline tests/test_streaming.py:129]: Can you add a one line statement what each condition is doing?

File: tests/test_streaming.py [hunk 3]
```diff
@@ -58,6 +59,94 @@ def test_dataloader_epoch_size_no_streams(local_remote_dir: Tuple[str,
...
+        assert samples_seen == (math.ceil(epoch_size / num_canonical_nodes) * num_canonical_nodes)
+        assert samples_seen_stream1 == int(
```
Comment (karan6181) [inline tests/test_streaming.py:131]: Can you also add a comment saying that stream1 data proportion is 40% and stream2 data proportion is 60% or something like that ? Thanks!

File: tests/test_streaming.py (hunk 1 above)
Comment (snarayan21) [reply tests/test_streaming.py:120]: haha yeah a bunch actually

File: tests/test_streaming.py (hunk 3 above)
Comment (snarayan21) [reply tests/test_streaming.py:131]: added as well!

**Review by karan6181 — APPROVED** [review]

**Pull request merged**